Add to swarm admin docs

mstanleyjones opened the pull request:
### Describe the proposed changes
- Add details about maintaining quorum
- Add details about pending tasks
### Project version

n/a
### Related issue

n/a
### Related issue or PR in another project

https://github.com/docker/docker/issues/25069#issuecomment-255218986
https://github.com/docker/docker/issues/26017#issuecomment-246405418
### Please take a look

@stevvooe @aaronlehmann 

<!-- To improve this template, edit .github/PULL_REQUEST_TEMPLATE.md. -->

Signed-off-by: Misty Stanley-Jones misty@docker.com

aaronlehmann commented on engine/swarm/admin_guide.md:
```diff
@@ -53,6 +53,29 @@ updates to the swarm. A quorum of managers must also agree on node additions
+### Maintaining a quorum of managers
+
+Always maintain **an odd number of manager nodes** in the swarm. Otherwise, your
+swarm may lose a quorum of managers, and will not be able to perform management
```
Having an even number of managers doesn't cause you to lose quorum, it just doesn't have any fault tolerance benefit over having one fewer manager. This is described in detail in the "Add manager nodes for fault tolerance" section below.

thaJeztah commented on engine/swarm/admin_guide.md:
```diff
@@ -53,6 +53,29 @@ updates to the swarm. A quorum of managers must also agree on node additions
+
+Always maintain **an odd number of manager nodes** in the swarm. Otherwise, your
+swarm may lose a quorum of managers, and will not be able to perform management
+tasks. Three managers is the minimum, and five managers is recommended. In a
```
Should we combine this with https://github.com/mstanleyjones/docker.github.io/blob/0db3e0bf877880a935d53b14459ee9bd09786656/engine/swarm/admin_guide.md#add-manager-nodes-for-fault-tolerance, as it's largely repeating the same message?

Note that having a single manager is "ok" (if no fault tolerance is needed at all)

aaronlehmann replied:
One manager is the minimum. Three managers is the minimum for a setup to continue working after a manager failure. I don't know if we should go so far as to recommend 5 managers, as 3 is sufficient for a lot of use cases. But it is good to point out that 5 managers allows you to lose 2.

thaJeztah commented on engine/swarm/admin_guide.md:
```diff
@@ -53,6 +53,29 @@ updates to the swarm. A quorum of managers must also agree on node additions
+from a manager node. This removes all managers except the manager the command
+was run from. Quorum is achieved because there is now only one manager. Promote
+nodes to be managers until you have the desired number of managers. You should
+always have an odd number of managers.
```
Should we point out that `--force-new-swarm` looses all swarm-related data (networks, services, etc)?

aaronlehmann replied:
I don't think we need to repeat the recommendation for an odd number of managers here. It's covered well under "Add manager nodes for fault tolerance".

aaronlehmann replied:
`--force-new-cluster` does not lose this data.

thaJeztah replied:
oh, I was misinformed then, I thought that was mentioned at some point. Thanks!

thaJeztah commented on engine/swarm/how-swarm-mode-works/services.md:
```diff
@@ -76,6 +76,27 @@ schedules tasks to worker nodes.
+You can even configure a service in such a way that no node currently in the
+swarm can run its tasks. Here are a few examples:
+
+- You can even create a service with an image that is not available in the
```
You can even -> You can?

thaJeztah commented on engine/swarm/how-swarm-mode-works/services.md:
```diff
@@ -76,6 +76,27 @@ schedules tasks to worker nodes.
+
+- You can even create a service with an image that is not available in the
+  registry, either because it hasn't been pushed to the registry yet or because
+  the registry is not currently available.
```
Using a non-existing image is actually dangerous, as (contrary to some other constraints), swarm does not know about this when scheduling, so it will keep trying to deploy the task, and pull images. It's possible the node runs out of resources because of this, or (if no log-rotation is in place) runs out of disk space.

aaronlehmann commented on engine/swarm/manage-nodes.md:
```diff
@@ -151,9 +151,11 @@ You can promote a worker node to the manager role. This is useful when a
+a node, you should always maintain **an odd number of manager nodes** in the
+swarm. Otherwise, your swarm may lose a quorum of managers, and will not be able
+to perform management tasks. For more information refer to the
+[Swarm administration guide](admin_guide.md).
```
As above, this warning feels a little dire compared to the actual consequences.

aaronlehmann commented on engine/swarm/manage-nodes.md:
```diff
@@ -199,6 +201,12 @@ quorum. To override the warning, pass the `--force` flag. If the last manager
+a node, you should always maintain **an odd number of manager nodes** in the
+swarm. Otherwise, your swarm may lose a quorum of managers, and will not be able
+to perform management tasks. For more information refer to the
+[Swarm administration guide](admin_guide.md).
```
Same comment about the severity of the warning.

aaronlehmann commented on engine/swarm/services.md:
```diff
@@ -128,15 +135,22 @@ deploys a service to the node. You can apply constraints to the
+`--reserve-memory` or `--reserve-cpu` flags. If no available nodes can satisfy
+the requirement (for instance, if you request 4 CPUs and no node in the swarm
+has 4 CPUs), the service remains in a pending state until a node is available to
+run its tasks.
```
Thanks for adding this. I think someone had pointed this out to me as something that needed more documentation, recently.

stevvooe commented on engine/swarm/how-swarm-mode-works/services.md:
```diff
@@ -76,6 +76,27 @@ schedules tasks to worker nodes.
+- You can impose placement constraints on the service, and the constraints may
+  not be able to be honored at a given time.
+
+This behavior is a consequence of the declarative way that Docker swarms are
```
Don't be a negatron. This wording makes it sound like a bad thing.

thaJeztah replied:
"a negatron" :joy:

mstanleyjones commented:
Addressed comments

aaronlehmann commented on engine/swarm/admin_guide.md:
```diff
@@ -53,6 +53,19 @@ updates to the swarm. A quorum of managers must also agree on node additions
 
+### Maintaining a quorum of managers
+
+If the swarm loses a quorum, it will not be able to perform management
```
Not sure what's most correct here, but I've generally seen "loses quorum" instead of "loses a quorum". "Losing the quorum of managers" might also work.

aaronlehmann commented on engine/swarm/admin_guide.md:
```diff
@@ -53,6 +53,19 @@ updates to the swarm. A quorum of managers must also agree on node additions
+### Maintaining a quorum of managers
+
+If the swarm loses a quorum, it will not be able to perform management
+tasks. If your swarm has multiple masters, always have more than two. In order
```
"managers", not masters

aaronlehmann commented on engine/swarm/admin_guide.md:
```diff
@@ -53,6 +53,19 @@ updates to the swarm. A quorum of managers must also agree on node additions
+to maintain a quorum, a majority of managers must be available. An odd number
+of managers is recommended, because the next even number does not make the quorum
+easier to keep. For instance, whether you have 3 or 4 managers, you can still only
+lose 1 manager and maintain a quorum. If you have 5 or 6 managers, you an still
```
Typo: `an` should be `can`

aaronlehmann commented on engine/swarm/admin_guide.md:
```diff
@@ -53,6 +53,19 @@ updates to the swarm. A quorum of managers must also agree on node additions
+only lose two.
+
+See [Recovering from losing a quorum](recovering-from-losing-a-quorum) for
+troubleshooting steps if you do lose a quorum of managers.
```
Same comment about "lose a quroum" / "losing a quorum", but again I'm not sure what's most correct.

aaronlehmann commented on engine/swarm/admin_guide.md:
```diff
@@ -239,27 +252,40 @@ you demote or remove a manager
+
+If you lose a quorum of managers, you cannot administer the swarm. Attempting
+to remove a node from a swarm which does not have a quorum of managers results
+in an error like the following:
```
This is technically true but I'm not sure why we would explain what happens when you try to remove a node from a swarm that already has too few nodes to function. Is it just an example of how management features stop working?

mstanleyjones commented:
All the feedback has been addressed (I believe). @aaronlehmann PTAL again

aaronlehmann commented on engine/swarm/admin_guide.md:
```diff
@@ -48,10 +48,22 @@ fault-tolerant. However, additional manager nodes reduce write performance
+of managers is recommended, because the next even number does not make the quorum
+easier to keep. For instance, whether you have 3 or 4 managers, you can still only
+lose 1 manager and maintain the quorum. If you have 5 or 6 managers, you can
+still only lose two.gn
```
Something wrong at the end of this paragraph.

johndmulhausen commented:
Deploy preview ready!

Built with commit 02ffe1f166dbbe70fe6b7f7dd7ac1e87fb494aff

https://deploy-preview-297--docker-docs.netlify.com

mstanleyjones commented:
@aaronlehmann I addressed your comments and also added info about forcing services to update. That would mean we need to hold back this PR until the 1.13 doc is merged in, so let me know if you want me to revert that rewording / addition and add it in a separate PR against `vnext-engine`.

aaronlehmann commented on engine/swarm/admin_guide.md:
```diff
@@ -231,29 +243,46 @@ you demote or remove a manager
-that existed before the loss of the quorum but it retains data necessary to the
-Swarm such as services, tasks and the list of worker nodes.
+When you run the `docker swarm init` command with the `--force-new-cluster` flag
+, the Docker Engine where you run the command becomes the manager node of a
```
comma on its own line

aaronlehmann replied:
Please fix the comma

aaronlehmann commented on engine/swarm/admin_guide.md:
```diff
@@ -267,11 +296,16 @@ balance across the swarm. When new tasks start, or when a node with running
-Use `docker service inspect --pretty <servicename>` to see the configured scale
+If you use an earlier version and you want to achieve an even balance of load
+across workers and don't mind disrupting running tasks, you can force your swarm
+to re-balance by temporarily scaling the service upward. Use
```
Hm, seems surprising to include workarounds for older versions in the documentation. Do we generally do this?

mstanleyjones replied:
I think we have to, since users of UCP and DTR won't be on 1.13 yet...

aaronlehmann replied:
Don't we version this document? I feel like it would be very confusing to try to cover multiple product versions in the same document.

mstanleyjones replied:
This is the admin guide that lives on docker.com. It is versioned in a way (you can get to older versions), but we also need to be careful because lots of people just Google. When a feature is brand new, we need to call out when you can expect it to work. Otherwise, support will get lots of calls from people on earlier versions trying to use commands.

mstanleyjones commented:
Feedback addressed. PTAL again.

aaronlehmann commented on engine/swarm/admin_guide.md:
```diff
@@ -231,29 +247,46 @@ you demote or remove a manager
+, the Docker Engine where you run the command becomes the manager node of a
+single-node swarm which is capable of managing and running services. The manager
+has all the previous information about services and tasks, but you need to add
+manager and worker nodes to achieve your previous task distribution and ensure
```
Workers would still be part of the swarm, but you would need to re-add managers.

aaronlehmann commented on engine/swarm/admin_guide.md:
```diff
@@ -267,11 +296,16 @@ balance across the swarm. When new tasks start, or when a node with running
+`docker service update` command to force the service to redistribute its tasks
+across the available worker nodes. This will cause the service tasks to restart.
+Client applications may be disrupted. If you have specified an `--update-delay`
+value for your service, the service will use a rolling restart.
```
See comment above. `--update-delay` does not control whether a rolling restart is used, only whether there is a delay between tasks being updated in a rolling update.

mstanleyjones commented:
Addressed latest feedback, including some comments I had missed before.

thaJeztah commented on engine/swarm/how-swarm-mode-works/services.md:
```diff
@@ -68,6 +68,31 @@ schedules tasks to worker nodes.
+  really have a node which can satisfy it.
+
+- You can impose placement constraints on the service, and the constraints may
+  not be able to be honored at a given time.
```
Reading this again; this section may be read as "instructions for creating a pending service". In some cases this is desirable (e.g. automatically start deploying when a node becomes available), however if this is just to prevent a service from being deployed, I think we should mention that scaling the service to `0` is the most logical way (or refer to the section about scaling).

thaJeztah commented on engine/swarm/manage-nodes.md:
```diff
@@ -6,10 +6,10 @@ title: Manage nodes in a swarm
 
 As part of the swarm management lifecycle, you may need to view or update a node as follows:
 
-* [list nodes in the swarm](manage-nodes.md#list-nodes)
```
did we fix this in Jekyll? Otherwise the `manage-nodes.md` is still needed :cry:

thaJeztah commented on engine/swarm/admin_guide.md:
```diff
@@ -231,29 +247,47 @@ you demote or remove a manager
+
+If you lose the quorum of managers, you cannot administer the swarm. If you have
+lost the quorum and you attempt to perform any management operation on the swarm,
+an error occurs:
```
@stevvooe you think we should repeat that tasks remain to run (https://github.com/docker/docker.github.io/pull/297/files#r88358814)? Train of thought; If the lost quorum is temporary, but all tasks are still running, there may not be a direct need to try to recover.

mstanleyjones commented:
Addressed the new feedback.
